ens-improvement-proposals/ensip-xx-canonical-name-resolver.md

```diff
@@ -0,0 +1,51 @@
+---
+description: Allows names to set a cname to alias records to another name
+---
+
+# ENSIP-XX: CNAME resolver
+
+| **Author**    | Jeff Lau \<jeff@ens.domains>
+| ------------- | ---------------------------------------------------------------- |
+| **Status**    | Draft                                                            |
+| **Submitted** | 2023-09-25                                                       |
+
+### Abstract
+
+Canonical Name records have existed for DNS since the 1987 in [RFC 1034](https://datatracker.ietf.org/doc/html/rfc1034). They allow aliasing a name to another name e.g. vitalik.eth to records.vitalik.eth. This specification outlines a resolver standard to implement CNAMEs at a resolver level without having to change the base ENS name resolution protocol.
+
+### Specification
+
+If a cname is found on a name, all records that are unset will be expected to proxy through the specified cname's records. Records that are already set on the resolver will ignore the cname.
+
+### Implementation
+
+```solidity
+interface ICnameResolver {
+    event CNameChanged(bytes32 indexed node, string name);
+    function cname(bytes32 node) external view returns (string memory);
```

Comment:
I think we need to make a decision here: Either cname takes a node and doesn't implement CCIP-Read and wildcards (so only onchain names can be CNAMEd to), or it takes a DNS-encoded name and does implement CCIP-Read and wildcards. The latter is the more future-proof option, but this implementation requirement needs to be called out.